oak/server/rust/oak_runtime/src/metrics/mod.rs

```diff
@@ -103,18 +125,92 @@ impl GrpcServerMetrics {
                 &["method_name"],
                 "Histogram of response latency of RPCs handled by the server.",
             )),
-            grpc_response_size_bytes: builder.register(histogram_vec(
-                "grpc_response_size_bytes",
+            grpc_server_response_size_bytes: builder.register(histogram_vec(
+                "grpc_server_response_size_bytes",
                 &["method_name"],
                 "Histogram of response sizes of RPCs handled by the server.",
             )),
             grpc_server_msg_sent_total: builder.register(histogram_vec(
                 "grpc_server_msg_sent_total",
                 &["method_name"],
-                "Total number of RPC response messages sent by the server.",
+                "Histogram of number of RPC response messages sent by the server.",
+            )),
+        }
+    }
+}
+
+impl GrpcClientMetrics {
+    fn new(builder: &MetricsBuilder) -> Self {
+        GrpcClientMetrics {
+            grpc_client_started_total: builder.register(counter_vec(
+                "grpc_client_started_total",
+                &["method_name"],
+                "Total number of RPCs started on the client.",
+            )),
+            grpc_client_completed: builder.register(counter_vec(
+                "grpc_client_completed",
+                &["method_name"],
+                "Total number of RPCs completed on the client, regardless of success or failure.",
+            )),
+            grpc_client_completed_latency_seconds: builder.register(histogram_vec(
+                "grpc_client_completed_latency_seconds",
+                &["method_name"],
+                "Histogram of rpc response latency for completed rpcs, in seconds.",
+            )),
+            grpc_client_sent_bytes: builder.register(histogram_vec(
+                "grpc_client_sent_bytes",
+                &["method_name"],
+                "Histogram of RPC messages sizes sent by the client.",
+            )),
+            grpc_client_received_bytes: builder.register(histogram_vec(
+                "grpc_client_received_bytes",
+                &["method_name"],
+                "Histogram of RPC messages sizes received by the client.",
+            )),
+            grpc_client_sent_msgs: builder.register(histogram_vec(
+                "grpc_client_sent_msgs",
+                &["method_name"],
+                "Histogram of number of RPC stream messages sent by the client.",
+            )),
+            grpc_client_received_msgs: builder.register(histogram_vec(
+                "grpc_client_received_msgs",
+                &["method_name"],
+                "Histogram of number of RPC stream messages received by the client.",
+            )),
+            grpc_client_failed_total: builder.register(counter_vec(
+                "grpc_client_failed_total",
+                &["failure_info"],
```

Comment:
For convenience, I'd probably go with 2) &ndash; we could maybe add a `method_name: Option<&str>` parameter to `send_error` and emit a default ("unknown"?) when it's `None`.

For the Roughtime thing, I think we can defer to a new Issue for now.  I do wonder if there's a possible refactoring of the `Invocation` class that would allow better encapsulation of the metrics stuff, in a way that makes re-use between Roughtime and gRPC client fall out naturally&hellip;